rust/candid/src/parser/token.rs

```diff
@@ -0,0 +1,277 @@
+use lalrpop_util::ParseError;
+use logos::{Lexer, Logos};
+
+#[derive(Logos, Debug, Clone, PartialEq, Eq)]
+pub enum Token {
+    #[regex(r"[ \t\r\n]+", logos::skip)]
+    // line comment
+    #[regex("//[^\n]*", logos::skip)]
+    #[token("/*")]
+    StartComment,
+    #[error]
+    UnexpectedToken,
+    #[token("=")]
+    Equals,
+    #[token("(")]
+    LParen,
+    #[token(")")]
+    RParen,
+    #[token("{")]
+    LBrace,
+    #[token("}")]
+    RBrace,
+    #[token(";")]
+    Semi,
+    #[token(",")]
+    Comma,
+    #[token(":")]
+    Colon,
+    #[token("->")]
+    Arrow,
+    #[token("null")]
+    Null,
+    #[token("vec")]
+    Vec,
+    #[token("record")]
+    Record,
+    #[token("variant")]
+    Variant,
+    #[token("func")]
+    Func,
+    #[token("service")]
+    Service,
+    #[token("oneway")]
+    Oneway,
+    #[token("query")]
+    Query,
+    #[token("blob")]
+    Blob,
+    #[token("type")]
+    Type,
+    #[token("import")]
+    Import,
+    #[token("opt")]
+    Opt,
+    #[token("==")]
+    TestEqual,
+    #[token("!=")]
+    NotEqual,
+    #[token("!:")]
+    NotDecode,
+    #[token("principal")]
+    Principal,
+    #[regex("[a-zA-Z_][a-zA-Z0-9_]*", |lex| lex.slice().to_string())]
+    Id(String),
+    #[token("\"")]
+    StartString,
+    // This token is not derived. Stores the unescaped string
+    Text(String),
+    #[regex("[+-]", |lex| lex.slice().chars().next())]
+    Sign(char),
+    #[regex("[0-9][_0-9]*", parse_number)]
+    Decimal(String),
+    #[regex("0[xX][0-9a-fA-F][_0-9a-fA-F]*", parse_number)]
+    Hex(String),
+    #[regex("[0-9]*\\.[0-9]*", parse_number)]
```

Comment:
This'll accept leading zeroes, do we want that?

Reply:
This is the same as the Motoko lexer, and the spec says it's should be the same as Motoko :)